File: benchmarks/Java/src/havlak/ControlFlowGraph.java
// Copyright 2011 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package havlak;

import som.Vector;


/**
 * A simple class simulating the concept of
 * a control flow graph.
 *
 * CFG maintains a list of nodes, plus a start node.
 * That's it.
 *
 * @author rhundt
 */
final class ControlFlowGraph {

  private final Vector<BasicBlock>  basicBlockMap;
  private BasicBlock                startNode;
  private final Vector<BasicBlockEdge> edgeList;

  ControlFlowGraph() {
    startNode = null;
    basicBlockMap = new Vector<>();
    edgeList = new Vector<>();
  }

  public BasicBlock createNode(final int name) {
    BasicBlock node;
    if (basicBlockMap.at(name) != null) {
      node = basicBlockMap.at(name);
    } else {
      node = new BasicBlock(name);
      basicBlockMap.atPut(name, node);
    }

    if (getNumNodes() == 1) {
      startNode = node;
    }
    return node;
  }

  public void addEdge(final BasicBlockEdge edge) {
    edgeList.append(edge);
  }

  public int getNumNodes() {
    return basicBlockMap.size();
  }

  public BasicBlock getStartBasicBlock() {
    return startNode;
  }

  public Vector<BasicBlock> getBasicBlocks() {
    return basicBlockMap;
  }
}

File: benchmarks/C++HJ/havlak/ControlFlowGraph.cpp
#include "ControlFlowGraph.h"

#include <iostream>

namespace havlak {

    ControlFlowGraph::ControlFlowGraph() {
        _startNode = nullptr;
        _basicBlockMap = make_shared<Vector<shared_ptr<BasicBlock>>>();
        _edgeList = make_shared<Vector<shared_ptr<BasicBlockEdge>>>();
    }

    shared_ptr<BasicBlock> ControlFlowGraph::createNode(int name) {
        shared_ptr<BasicBlock> node;
        if (_basicBlockMap != nullptr && _basicBlockMap->atPtr(name) != nullptr) {
            node = _basicBlockMap->atPtr(name);
        } else {
            node = make_shared<BasicBlock>(name);
            _basicBlockMap->atPut(name, node);
        }

        if (getNumNodes() == 1) {
            _startNode = node;
        }
        return node;
    }

    void ControlFlowGraph::addEdge(shared_ptr<BasicBlockEdge> edge) {
        _edgeList->append(edge);
    }

    int ControlFlowGraph::getNumNodes() {
        return _basicBlockMap->size();
    }

    shared_ptr<BasicBlock> ControlFlowGraph::getStartBasicBlock() {
        return _startNode;
    }

    shared_ptr<Vector<shared_ptr<BasicBlock>>> ControlFlowGraph::getBasicBlocks() {
        return _basicBlockMap;
    }

}